libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,648 @@
+#include <__ranges/view_interface.h>
+#include <__type_traits/conditional.h>
+#include <__type_traits/decay.h>
+#include <__type_traits/is_nothrow_constructible.h>
+#include <__type_traits/make_unsigned.h>
+#include <__type_traits/maybe_const.h>
+#include <__utility/forward.h>
+#include <__utility/in_place.h>
+#include <__utility/move.h>
+#include <tuple>
+#include <variant>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_PUSH_MACROS
+#include <__undef_macros>
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 26
+
+namespace ranges {
+
+#  ifdef __cpp_pack_indexing
+template <class... _Tp>
+using __extract_last _LIBCPP_NODEBUG = _Tp...[sizeof...(_Tp) - 1];
+#  else
+template <class _Tp, class... _Tail>
+struct __extract_last_impl : __extract_last_impl<_Tail...> {};
+template <class _Tp>
+struct __extract_last_impl<_Tp> {
+  using type _LIBCPP_NODEBUG = _Tp;
+};
+
+template <class... _Tp>
+using __extract_last _LIBCPP_NODEBUG = __extract_last_impl<_Tp...>::type;
+#  endif
+
+template <bool _Const, class... _Tp>
+struct __all_but_first_model_sized_range;
+
+template <bool _Const, class _Head, class... _Tail>
+struct __all_but_first_model_sized_range<_Const, _Head, _Tail...> {
+  static constexpr bool value = (sized_range<__maybe_const<_Const, _Tail>> && ...);
+};
+
+template <bool _Const, class... _Views>
+concept __all_random_access = (random_access_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class... _Views>
+concept __all_bidirectional = (bidirectional_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class... _Views>
+concept __all_forward = (forward_range<__maybe_const<_Const, _Views>> && ...);
+
+template <bool _Const, class _First, class... _Tail>
+struct __all_common_ignore_last {
+  static constexpr bool value =
+      common_range<__maybe_const<_Const, _First>> && __all_common_ignore_last<_Const, _Tail...>::value;
+};
+
+template <bool _Const, class _Tail>
+struct __all_common_ignore_last<_Const, _Tail> {
+  static constexpr bool value = true;
+};
+
+template <bool _Const, class... _Rs>
+concept __concat_is_random_access =
+    (__all_random_access<_Const, _Rs...>) && (__all_common_ignore_last<_Const, _Rs...>::value);
+
+template <bool _Const, class... _Rs>
+concept __concat_is_bidirectional =
+    (__all_bidirectional<_Const, _Rs...>) && (__all_common_ignore_last<_Const, _Rs...>::value);
+
+template <input_range... _Views>
+  requires((view<_Views> && ...) && (sizeof...(_Views) > 0) && __concatable<_Views...>)
+class concat_view : public view_interface<concat_view<_Views...>> {
+  tuple<_Views...> __views_;
+
+  template <bool _Const>
+  class __iterator;
+
+public:
+  _LIBCPP_HIDE_FROM_ABI constexpr concat_view() = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit concat_view(_Views... __views) : __views_(std::move(__views)...) {}
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator<false> begin()
```

Comment:
```suggestion
  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr __iterator<false> begin()
```
Please annotate sccording to: https://libcxx.llvm.org/CodingGuidelines.html#apply-nodiscard-where-relevant + nodiscard.verify.cpp test in the relevant subdirectory tests/libcxx